Wandb resume ddp

AyushExel opened the pull request.

AyushExel commented:
Update:
* PR merged for limiting predictions to top 10 when plotting -> https://github.com/AyushExel/yolov5/pull/2#issuecomment-1183934486

@Laughing-q I think there's CUDA memory leak somewhere.. As you can see in the GPU memory allocation graph(brown line), it increases progressively.. It should generally be constant like the other 2 lines that represent other devices being used by yolov5 master branch. I'll investigate today. Maybe fixing that will also improve the speed and not freeze the system
<img width="644" alt="Screenshot 2022-07-14 at 9 06 20 AM" src="https://user-images.githubusercontent.com/15766192/178892609-997b23bd-f412-4dd7-804f-930308ce4300.png">

EDIT: It goes up and down quite a bit.. One thing could be the transference of masks from cuda to cpu but that should only occur after epoch finishes..This is only for the first epoch.

AyushExel commented:
@Laughing-q It froze again. Now i see what's causing the system to freeze. System RAM is being taken up ..there's definitely memory leak not just CUDA ..See this..It keeps increasing until there's no more left which causes system to freeze
<img width="678" alt="Screenshot 2022-07-14 at 9 42 49 AM" src="https://user-images.githubusercontent.com/15766192/178897113-50e6c1ac-39b6-4dad-96c9-84728356db06.png">

Laughing-q commented:
@AyushExel  I wonder if you can set `plots=False` to see if there is still this memory leak problem. If  there no memory leak, we can check the relevant code of `plots`.

AyushExel commented:
@Laughing-q yes doing exactly that right now -- [Run here]( https://wandb.ai/cayush/YOLOv5-Inst-seg/runs/1q2zri45?workspace=user-cayush)

Laughing-q commented:
@AyushExel I did some test in a small dataset, seems this upsample increases cuda memory.
[https://github.com/AyushExel/yolov5/blob/529c5401d5c2b371d6909c1d4c874f37d41b2ff1/utils/segment.py#L183](https://github.com/AyushExel/yolov5/blob/529c5401d5c2b371d6909c1d4c874f37d41b2ff1/utils/segment.py#L183)
But I didn't reproduce the RAM memory issue.

AyushExel commented:
Moving our discussion here..
<img width="674" alt="Screenshot 2022-07-14 at 7 40 01 PM" src="https://user-images.githubusercontent.com/15766192/179017054-27dd90c4-3bd7-4bb1-a056-a36212c46dff.png">

@Laughing-q I think we should focus on fixing the memory leak in evaluator. See how training for 1 epoch takes only 20GB which same as training detection..but It jumps by double to 45GB during evaluation.. There is no reason for this much increase in consumption of memory.. first, because val set is m much smaller. Second, there is no augmentation in val set, thirdly, the val loader is already loaded before training so it should not take up any more memory.. SO there is no reason for this much memory cost.. it we can fix this, I think plots will not be an issue..
What do you think?

Laughing-q commented:
@AyushExel yes I think so, that 's so strange. Could you tell me more about your experiment settings? I'll do some tests tomorrow to reproduce your results. Maybe my test datasets were too small, I didn't see a huge increase in system memory. I'll try COCO tomorrow.

AyushExel commented:
@Laughing-q I'm using the same branch (AyushExel/instance_seg). You can do git pull to be up to date.. 
My command is - `python train_instseg.py --cfg yolov5s_seg.yaml --data coco.yaml --device 7 --workers 8 --weights --project yolov5-inst-seg --batch 32 --noplots
`

Laughing-q commented:
@AyushExel About [https://github.com/AyushExel/yolov5/pull/4#issuecomment-1184549461](https://github.com/AyushExel/yolov5/pull/4#issuecomment-1184549461)
I agree with the lower accuracies when set `mask_ratio=4`, I tested on my own small dataset, and the accuracy was about the same as mask-ratio=1, but it could save me a lot of CUDA memory, so I didn't have many tests, just set mask-ratio=4. Maybe you can set mask-ratio=4 for a test to see the cuda memory and system memory.

I also agree that you can turn off mosaic for a test, disabling this would save some cuda memory in my previous experiments.
Another way is to cancel the offset in yolov5's label assign, just use the center grid, but this might get lower accuracies.

In fact, I really wanted to optimize the loss function when I was doing this project. But the current method is modified according to yolact, when calculating mask loss, it can only be calculated according to the dimension of batch which was different with yolov5(calculated loss in multi-level feature dimension), so the loss needs to be calculated circularly in two different dimensions. I'll work on this after we solve the memory issue, maybe we can change a method, not yolact.

AyushExel commented:
@Laughing-q Ohh yes we should definitely change the loss calculation method from yolact to yolov5.. I think we should do things based on solov2.. Everything from segmentation head, loss, etc..because it had much better results than yolact

Laughing-q commented:
@Hey man, I just confirmed that segment loss calculation is based on downsample ratio 4, same as yolact. 
[https://github.com/AyushExel/yolov5/blob/c656a5771f1746874be6d4617efa0abb3901d0c6/utils/seg_loss.py#L145](https://github.com/AyushExel/yolov5/blob/c656a5771f1746874be6d4617efa0abb3901d0c6/utils/seg_loss.py#L145)
So  the effect of mask ratio=4 or 1 is the same about the accuracy, the difference is that one is to downsample in advance and then put it into cuda which save a lot memory(both system and cuda), and the other is to put the original size into cuda then downsample before calculating the loss. 

By the way, I didn't reproduce your huge memory growth, I used the same command with yours. reset the batch-size to 8 cause I don't have that much cuda memory.

AyushExel commented:
@Laughing-q that's strange.. did you use my branch ?

Laughing-q commented:
@AyushExel yes, `instance_seg` with the latest code, the only difference is batch-size=8.

AyushExel commented:
@Laughing-q ok.. I'll test it with batch size 8 again today. Have you looked into solov2 paper or code?

Laughing-q commented:
@AyushExel Yes, I've seen the paper and code before, but I forgot some details. I need to review them.

AyushExel commented:
@Laughing-q ok..I think we should use concepts from solov2 because it outperforms yolact in both speed and accuracy.
By the way, can you share the dashboard link to the run that you ran? I want to replicate the memory result from there

AyushExel commented:
@Laughing-q so I started training with mask-ratio 4 and train batches were not plotted.. I think in your last PR you updated it so that mask ratio and plots=True are not coupled right? Or maybe you did that just for preds??

Laughing-q commented:
@AyushExel yes, just preds, it won't plot training images, cause you set None for the last param here:
[https://github.com/AyushExel/yolov5/blob/04f78b55bdaa5b0fd6b8d6d11cba6337410e6a1c/train_instseg.py#L412](https://github.com/AyushExel/yolov5/blob/04f78b55bdaa5b0fd6b8d6d11cba6337410e6a1c/train_instseg.py#L412)

Laughing-q commented:
> @Laughing-q ok..I think we should use concepts from solov2 because it outperforms yolact in both speed and accuracy. By the way, can you share the dashboard link to the run that you ran? I want to replicate the memory result from there

[https://wandb.ai/laughing/runs/runs/1tcv20es/system](https://wandb.ai/laughing/runs/runs/1tcv20es/system)
but I found a strange thing, the results of recording memory changes with locally written shell scripts are inconsistent with those on wandb. Here is my local result:
![plot](https://user-images.githubusercontent.com/61612323/179152279-e5102cb8-8145-4b52-9efb-d7b244712fd5.jpg)
The decline indicates the evaluation stage. (I count the total memory, so it may look a little higher, but it seems to be in line with expectations), I also observed it locally before, so I didn't find any memory exceptions.

and the command I used is
```shell
free --mebi | sed -n '2{p;q}' | awk '{printf ("%2.2f\n", ( $3 / 1024))}'
```

AyushExel commented:
@Laughing-q Ok..Let's not worry about the memory issue then..I'm training with mask-ratio 4 and it doesn't take that much memory..so maybe spikes won't be a problem.. Once this benchmark is done.. Let's focus more on optmizing the loss function and move from yolact code to yolov5 as it is much faster. In the mean-time I'll read the solov2 paper to get some inspiration for segmentation heads

AyushExel commented:
@Laughing-q OK so my initial thoughts about downsampling(mask-ratio 4) is that is does affect accuracies, at least in intial epochs I'm seeing the models with mask-ratio 1 were performing much better. But yeah let's think about this later. Let's just focus on getting the loss computation fast.

AyushExel commented:
<img width="671" alt="Screenshot 2022-07-15 at 12 58 42 PM" src="https://user-images.githubusercontent.com/15766192/179174204-7e478b4d-9b5d-434b-b4e5-3536163aaa17.png">
Ok so leakage is still there.. But the system isn't freezing because mask-ratio 4 takes less memory.. If this keeps increasing and gets to 52GB, it'll definitely crash..

Laughing-q commented:
> @Laughing-q OK so my initial thoughts about downsampling(mask-ratio 4) is that is does affect accuracies, at least in intial epochs I'm seeing the models with mask-ratio 1 were performing much better. But yeah let's think about this later. Let's just focus on getting the loss computation fast.

I believe I probably know the reason for this, let's talk about this later.

Laughing-q commented:
@AyushExel I looked at the code of solov2. It also needs to cycle twice in the batch and multi-level dimensions to calculate loss.
I add some convolution operation to the current head according to the head of solov2. I will test it to see the result. By the way, there are some bugs in the DDP mode of the current code which we may fix it later. I temporarily used yolov5-q to do the test.

AyushExel commented:
@Laughing-q can you tell me about the DDP bug I'll take a look once this training finishes? Also do you see an option to open another issue on this repo? I don't see it but I'd like to take this conversation in another thread and keep this PR only for major improvement updates

Laughing-q commented:
@AyushExel I didn't see the option either. But ddp issue is probably caused by the code merging from `yolov5-q`, no big deal. You can test it after you finish training then you'll know, or I will fix it later and commit a new PR.

AyushExel commented:
@Laughing-q [Here's]( https://wandb.ai/cayush/yolov5-instseg/runs/3d38akrm) the latest run, I'll let it run overnight. 
One thing that seems very odd is the mAP of both boxes and masks @.95 .. It seems too low when compared to mAP @ .5..maybe there could be some bug in calculation but otherwise the results seem good.

tomorrow, I'll try to understand and optimize the loss function

AyushExel commented:
@Laughing-q hey feel free to make conflicting PRs.. I'll. resolve the conflicts and merge them :)

AyushExel commented:
@Laughing-q looks like the latest change is breaking the original train.py. Look at the tests

Laughing-q commented:
@AyushExel Ooh, I see, this update is breaking the detection training, which I haven't test. but it won't break segmentation training.

Laughing-q commented:
@AyushExel so should I make a new PR to fix this? or just let it be for now.

AyushExel commented:
@Laughing-q no problem for now! Let's focus on segmentation only.. This can be left for later when we merge

AyushExel commented:
@Laughing-q I added a change in wandblogger to match steps.. Now 300 epochs should create 300 steps..Removed the logging of train labels as it was not important

Laughing-q commented:
@AyushExel okay, great!

AyushExel commented:
@Laughing-q I'm training the m model with all the latest updated using the `hyp-med.yaml` from scratch. 
[Link](https://wandb.ai/cayush/yolov5-seg-newlabels?workspace=user-cayush)

AyushExel commented:
@Laughing-q this branch is now compatible with torch 1.12 and later...1.12.0 has some GPU issues so I'll use 1.13.0 nightly for training so the results are deterministic.

AyushExel commented:
@Laughing-q uh oh :( 
```
RuntimeError: upsample_bilinear2d_backward_out_cuda does not have a deterministic implementation, 
but you set 'torch.use_deterministic_algorithms(True)'. You can turn off determinism just for this operation, 
or you can use the 'warn_only=True' option, if that's acceptable for your application. 
```
Can we just use `F.interpolate(scale_factor=2)` instead of nn.Upsample ?

AyushExel commented:
@Laughing-q in my branch I'm using this for enabling deterministic behaviour. I'm not sure if it'll harm the model. let's see
https://github.com/AyushExel/yolov5/commit/12522b5cb3c9c1d7b235ee049bd478ded4a79295

AyushExel commented:
@Laughing-q I added the changes you suggested. Can you review the last commit?

Laughing-q commented:
@AyushExel yes I think there's no problem here.

AyushExel commented:
@Laughing-q ok bilinear mode does not support deterministic behaviour.. I've switched to default(nearest) mode in my experimental branch

Laughing-q commented:
@AyushExel okay

Seperendity commented:
@AyushExel Thank you for your excellent work, I would like to ask you if the total loss function here is made up of the loss of detection and the loss of instance segmentation together, I didn't find the exact composition of the loss, can you tell me where the part of the code? Will the learning of segmentation branch help detection task?

AyushExel commented:
@Seperendity the total loss just combines the box and segmentation (pixel) loss. You can find the details in computeLoss class
